integrations/llama_cpp/src/haystack_integrations/components/generators/llama_cpp/chat/chat_generator.py

```diff
@@ -254,9 +304,95 @@ def run(
         for choice in response["choices"]:
             chat_message = self._convert_chat_completion_choice_to_chat_message(choice, response)
             replies.append(chat_message)
-
         return {"replies": replies}
 
+    @staticmethod
+    def _handle_streaming_response(
+        response_stream: Iterator[CreateChatCompletionStreamResponse],
+        streaming_callback: SyncStreamingCallbackT,
+        component_info: ComponentInfo,
+    ) -> Dict[str, List[ChatMessage]]:
+        """
+        Take streaming responses from llama.cpp, convert to Haystack StreamingChunk objects, stream them,
+        and finally convert them to a ChatMessage.
+
+        :param response_stream: The streaming responses from llama.cpp.
+        :param streaming_callback: The callback function for streaming chunks.
+        :param component_info: The component info.
+        :returns: A dictionary with the replies.
+        """
+        streaming_chunks = []
+
+        seen_tool_call_ids = set()  # Track tool call IDs we've seen
+
+        for i, chunk in enumerate(response_stream):
+            content = ""
+            tool_calls = []
+            mapped_finish_reason = None
+
+            # Track new tool call IDs in this chunk.
+            # Considering tool call ID is the only reliable way to recognize tool calls in llama.cpp streaming.
+            # They are often spread across multiple chunks.
+            new_tool_call_ids = set()
+
+            if chunk.get("choices"):
+                choice = chunk["choices"][0]
```

Comment:
Can it happen to have 0-length `choices`?

Reply:
Should not happen but I now adopted a more defensive approach.

Reply:
@anakin87 should we open an issue to update this in our other chat generators?

Reply:
Are you referring to `choices` length? I think it depends on how this part is implemented in different Chat Generators...

Reply:
This seems like a good idea and it's also a "good first issue". I'll create an issue.